Add C8 REST guidelines (#4580)
camunda/camunda-docs: docs/apis-tools/camunda-api-rest/camunda-api-rest-guidelines.md

---
id: camunda-api-rest-guidelines
title: "Guidelines"
description: "Learn about the basic guidelines, structures, and conventions of the Camunda 8 REST API."
---

> Camunda follows a mix of proposed standards and best practices for RESTful design and consistent implementation across all components. This ensures customers working across all component APIs have a consistent and expected experience without having to study our API reference material or perform “validation testing” to see how our APIs respond.

Reviewer: ```suggestion
Camunda follows a mix of proposed standards and best practices for RESTful design and consistent implementation across all components. 
```

> ## Naming conventions

Reviewer: This section should be rewritten for a non-Camundi audience or consumers of the API, not builders of the API.

> Naming should be simple, intuitive, and consistent across Camunda 8 APIs. This improves the experience when working across multiple APIs by reducing friction.

Reviewer: ```suggestion

```

I would remove these sentences or change the tense. "Naming is simple, intuitive, ..."

When working with IDs or similar short 2- or 3-letter words or acronyms, don’t capitalize all letters, only capitalize the first letter. If standalone, all letters should be lowercase.

| term | usage                                      |
| ---- | ------------------------------------------ |
| ID   | _id_ (standalone) or _processDefinitionId_ |
| URL  | _url_ (standalone) or _externalUrl_        |
| UUID | _uuid_ (standalone) or _clusterUuid_       |

Identifiers follow a naming rule, in parameters and data attributes alike:

- Unique technical identifiers are suffixed with **key**, for example, _userTaskKey_, _processInstanceKey_, _userKey_. They are numeric values in most cases.
- Other identifiers, such as copied identifiers from the BPMN XML, can be arbitrarily named but are usually suffixed with **id**, for example, _processDefinitionId_.

## Versioning

Camunda uses the term “major version number” from [semantic versioning](https://semver.org/), but do not follow semantic versioning for APIs outright. Instead, Camunda provides updates to the API in place and only increment the version number when a major, breaking change happens.

> Adding attributes and endpoints are not considered breaking changes. Breaking changes can potentially break an integration. See [GitHub’s REST documentation](https://docs.github.com/en/rest/about-the-rest-api/breaking-changes?apiVersion=2022-11-28#about-breaking-changes-in-the-rest-api) for a comprehensive summary of breaking changes. No migration is required unless there is a breaking change.

Reviewer: ```suggestion
:::note 
New attributes and endpoints are not considered breaking changes.
:::

Camunda avoids breaking changes that may break integrations or require migrations. Camunda follows [GitHub’s REST documentation](https://docs.github.com/en/rest/about-the-rest-api/breaking-changes?apiVersion=2022-11-28#about-breaking-changes-in-the-rest-api) for a comprehensive summary of breaking changes. 
```

Feel free to further tweak this, but this is the general idea I'd like to see.

The API version number does not match the product version (8.x.x). An API’s version is rather defined by the API version number and the product version, for example, “_POST /v2/user-tasks/search_ in Camunda 8.7.0”.

Camunda does API versioning rather than endpoint versioning, for example, the version changes for all endpoints if there is a breaking change in at least one endpoint. Multiple versions of an API can exist in a product version to allow for a migration period, for example, “POST /v2/user-tasks/search and POST /v3/user-tasks/search in Camunda 8.7.0”.

## HTTP status codes & error handling

> Handling errors must be consistent across all endpoints, using well-known HTTP status codes and clear descriptions. This includes the information about errors and the use of a problem details object.

Reviewer: "Handling errors is consistent..."

```
{
  "items": [
    { <field1>: <value1>, <field2>: ... }
    { <field1>: <value1>, <field2>: ... }
  ],
  "page": {
    "totalItems": <total  hits>,
    "firstSortValues": [
      <sort field1 value>,
      <sort field2 value>,
      ... ,
      <internal tiebreaker value>
    ],
    "lastSortValues": [
      <sort field1 value>,
      <sort field2 value>,
      ... ,
      <internal tiebreaker value>
    ]
  }
}
```

### Search example

Querying for the first 3 user tasks with certain criteria, sorted by state, could look as follows:

```
POST /v2/user-tasks/search

{

>   "filter": {
>     "assignee": {"$eq": "demo"},
>     "candidateGroups": { "$in": ["groupA", "groupB"] },
>     "variables" : {
>       "orderVolume": 10000,
>       "foo": {"$lt": 500},
>       "bar": {"$exists": false}
>     },
>     "$or": [
>       { "priority": {"$eq": "high"}, "dueDate": { "$lt": "<date>" } },
>       { "priority": {"$eq": "low"}, "followUpDate": { "$lt": "<date>" } }
>     ]

Reviewer 2: 🔧 No `$or` yet, add multiple conditions.

```suggestion
  "filter": {
    "assignee": {"$eq": "demo"},
    "candidateGroups": { "$in": ["groupA", "groupB"] },
    "variables" : {
      "orderVolume": 10000,
      "foo": {"$gt": 100, "$lt": 500},
      "bar": {"$exists": false}
    }
```

We follow the proposed standard from [RFC 9457](https://www.rfc-editor.org/rfc/rfc9457) for problem details. The problem object contains at least the following members:

- Type
- Status
- Title
- Detail
- Instance

We use the following error codes and descriptions across our APIs:

- 200 OK
- 204 No Content
- 400 Bad Request
  - Generic error that contains further description in the problem detail.
- 401 Unauthorized
  - The client is not authenticated yet.
  - The client should try again with a modified Authorization header.
- 403 Forbidden
  - The client has incorrect or insufficient permissions for the request.
- 404 Not Found
- 409 Conflict
  - The request is trying to modify a resource that is currently not in the right state.
- 412 Precondition failed
  - The client should check the cluster status.
- 429 Rate Limited Exceeded
  - The client exceeds a defined limit of requests, e.g. Zeebe signaling backpressure due to more requests than the broker can currently process
- 500 Internal Server Error

>   - Generic error that contains further description in the problem detail.

Reviewer: I would recommend turning this into a table.

Search endpoints can also be used to directly access entity instances with a unique identifier. As an alternative, resources can also provide GET method endpoints for fetching the data of single instances. In most cases, this is done by a specific key parameter in the URL path, e.g. _GET /v2/users/{userKey}_.

### Search requests

Query requests consist of the components for **filter**, **sort**, and **page**.

The **filter** object defines which fields should match. Only items that match the given fields will be returned. The available fields vary by object and are described in the respective search endpoint. Filtering by a unique identifier is usually available in filtering options. Beyond that, the filter options don’t have to comprise all the returned items’ attributes.

The **sort** object specifies which fields of the object should be sorted and whether they are sorted in ascending (ASC) or descending (DESC) order.

> The **page** object details how to slice the result set. An initial search request can omit the page object or define the _limit_. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned _firstItemSortValues_ and _lastItemSortValues_ of the [search responses](#search-responses) to page through the items by copying that array into one of the attributes _searchAfter_ or _searchBefore_.

Reviewer 3: This is really helpful, actually...I was going to ask you a question about these fields today anyway. 👍

Reviewer 3: @tmetzke stopping back after attempting to add spec descriptions for these properties, armed with my new understanding -- one other field that isn't mentioned here is `from`, an int32. 

1. Am I correct in thinking this is probably an index to start from? e.g. `from: 11` means start at the 11th item. 
2. If I am correct, is that a zero-based index or 1-based?
3. Do you think we should describe that here, since we covered the other fields?

Reviewer 3: Potentially blocking: 

Since these are fields, I think they should be `code` instead of _italic_. @christinaausley can probably confirm, but that's how I interpret the [code blocks section of the style guide](https://confluence.camunda.com/pages/viewpage.action?spaceKey=HAN&title=Camunda+style+guide).

```suggestion
The **page** object details how to slice the result set. An initial search request can omit the page object or define the `limit`. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned `firstItemSortValues` and `lastItemSortValues` of the [search responses](#search-responses) to page through the items by copying that array into one of the attributes `searchAfter` or `searchBefore`.
```

Reviewer 3: Also blocking:

The [properties in the spec](https://github.com/camunda/camunda/blob/main/zeebe/gateway-protocol/src/main/proto/rest-api.yaml#L4345-L4349) are named `firstSortValues` and `lastSortValues`.

```suggestion
The **page** object details how to slice the result set. An initial search request can omit the page object or define the _limit_. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned _firstSortValues_ and _lastSortValues_ of the [search responses](#search-responses) to page through the items by copying that array into one of the attributes _searchAfter_ or _searchBefore_.
```

```
{
  "filter":  {
    <entity-specific  filter  options>
  },
  "sort":  [
    {  "<field>":  "<order>"  },
    {  "<field>":  "<order>"  }
  ],
  "page":  {
    "searchAfter":  [  …  ,  …  ],
    "searchBefore":  [  …  ,  …  ],
    "limit":  <limit>
  }
}
```

#### Advanced Search

To provide an easy yet expressive way for users to search for and filter resources, search requests can contain more advanced filter criteria than fields being _equal_ to a target value.

For example, this allows searching using logical (and, in) and comparison operators (greater than, less than). The list of supported advanced filter operators is described below. Examples of advanced operators are shown in the [Search example](#search-example).

##### Conditional operators

| Operator | Syntax                                    | Description                                                                                                                                               |
| -------- | ----------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | $eq      | field: { "$eq": value }                   | Filter where field is equal to value. Abbreviated form field: value is also allowed.                                                                      |

Reviewer 3: ```suggestion
| `$eq`      | `field: { "$eq": value }`                   | Filter where field is equal to value. Abbreviated form field: value is also allowed.                                                                      |
```

Potentially blocking:

Again, this is my interpretation of the style guide, but I think these bits should all be inline code. (Applies also to many lines that follow this one.)

### Search requests

Query requests consist of the components for **filter**, **sort**, and **page**.

<Tabs groupId="requestComponents" defaultValue="filter" queryString values={[{label: 'Filter', value: 'filter', },{label: 'Sort', value: 'sort', },{label: 'Page', value: 'page', },]} >
<TabItem value="filter">
The filter object defines which fields should match. Only items that match the given fields will be returned. The available fields vary by object and are described in the respective search endpoint. Filtering by a unique identifier is usually available in filtering options. Beyond that, the filter options don’t have to comprise all the returned items’ attributes.
</TabItem>
<TabItem value="sort">
The sort object specifies which fields of the object should be sorted and whether they are sorted in ascending (ASC) or descending (DESC) order.
</TabItem>
<TabItem value="page">

> The page object details how to slice the result set. An initial search request can omit the page object or define the limit. This specifies the maximum number of results to retrieve per request. Subsequent requests can use the value of the returned firstItemSortValues and lastItemSortValues of the [search responses](#search-responses) to page through the items by copying that array into one of the attributes searchAfter or searchBefore.

Reviewer 4: Did we ever figure out a workaround for linking within tabs?

Reviewer: @christinaausley do you mean linking from one tab to another? You might be able to use a [query string](https://docusaurus.io/docs/markdown-features/tabs#query-string), assuming the tabs are in the same group (since it's not an anchor, the browser won't scroll).

Reviewer 4: As in, can we link to another page from within a tab. Nevermind though, as I realized this just looked wonky on my local build and looks fine on deploy 👍
